File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
      let sorted = nums.sort((a, b) => a - b)
  let res = []
  
  for(let i = 0; i <= sorted.length; i++){
    let targ = sorted[i]
    if(i > 0 && targ === sorted[i - 1])
      continue
    
  	let [l, r] = [i + 1, sorted.length - 1]
    
    while (l < r){
      	let sum = targ + sorted[l] + sorted[r]
    	if(sum < 0 ) l+=1
      	else if(sum > 0 ) r-=1
      	else { 
          res.push([targ, sorted[l], sorted[r]]); 
          l += 1 
          while(nums[l] === nums[l - 1] && 1 < r){
          	l+=1
          } 
        }
      	
    }
  }
  
  return res
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        sorted_nums = sorted(nums) # [-4,-1,-1,0,1,2]
        res = [];
        for i in range(len(sorted_nums)):
            
            if i>0 and sorted_nums[i-1]==sorted_nums[i]:
                continue
                
            [left, right] = [i + 1, len(sorted_nums) -1 ]
            while left < right:
                total = sorted_nums[i] + sorted_nums[left] + sorted_nums[right]
                if total < 0:
                    left += 1
                elif total > 0:
                    right -= 1
                else:
                    res.append([sorted_nums[i], sorted_nums[left], sorted_nums[right]])
                    left += 1
                    while sorted_nums[left - 1] == sorted_nums[left] and left < right:
                        left += 1
        
        return res